### ISSUE TITLE
bundle built-in crypto module

### ISSUE THREAD

**Issue by jasonswearingen** [substack/node-browserify#1467]
I can not figure out how to create a shared/vendor bundle that includes the crypto-browserify module so that it's not automatically included in my app bundle. (shaving 60kb or so off the size)

I tried adding
```
 "browser": {
    "crypto": "crypto-browserify"
  }
```
to my ```package.json``` but that doesn't work

What seems to work is if I add ```crypto``` to my package.json, but ```crypto``` is ***NOT*** the module browserify uses (it seems it's some other javascript implementation)

how can I add the ```crypto-browserify``` to my shared bundle properly?

**Comment by terinjokes** [comment]
Assuming you're running a recent version of browserify, you don't have to
modify the browser field at all, `require('crypto')` will automatically get
`crypto-browserify`

**Comment by jasonswearingen** [comment]
yes it does, the problem is that it's always put in my app.bundle.js file, which adds 60kb to it.

I'd like to put ```crypto``` (and all the builtins actually, but crypto is the biggest) into a ```vendor-shared.bundle.js``` file, but can not figure how to do this.  It's always added to my ```app.bundle.js``` file.

**Comment by jasonswearingen** [comment]
fyi, i'm following the multiple-bundles tutorial here: http://fettblog.eu/gulp-browserify-multiple-bundles/

**Comment by terinjokes** [comment]
It's always put into your bundle because its required by your code.

You can try setting crypto to "false" in your package.json and bundling it
separately.

**Comment by jasonswearingen** [comment]
Thanks for the tip, but unfortunately it has no effect.  Here's what I added to my ```package.json```
```
  "browser": {
    "crypto": false
  }
```
The only impact i have is if I include the actual ```npm crypto``` module.  That seems to override the browserify auto-bundling of ```crypto-browserify```.  Otherwise no matter what I try, ```crypto-browserify is always added to my bundle.  

If I turn off all built-ins, ```crypto-browserify``` isn't bundled of course, but it never finds a crypto module to use, even if I add ```crypto-browserify``` explicitly in my ```package.json```

**Comment by jasonswearingen** [comment]
maybe the issue is that my app.js file doesn't actually reference the ```crypto``` module.  it's a nested dependency by some other module I'm loading.   So these "browser field" settings don't seem to work.

**Comment by jasonswearingen** [comment]
*sigh* nope, that's not it.  I explicity include ```crypto``` and it's still getting loaded up.

I'd venture to guess that multiple-bundling of the builtins just doesn't work.  probably not ever considered to be tested :(

**Comment by jasonswearingen** [comment]
my hacky workaround:

in my gulp file:
1) when creating the shared bundle, I rename ```crypto-browserify``` to ```crypto``` via the ```b.require(path,{expose:id})``` function.  
2) when creating my app bundle, I set ```crypto``` as an external.

it works, but obviously not ideal.

**Comment by terinjokes** [comment]
If you can create a gist using just the browserify API of what you wanted
to do, we might have a better idea of what to consider changing.

Sorry it isn't working out of the box right now.

**Comment by jasonswearingen** [comment]
something like this, hope it helps.   this is just the normal creating 2 bundles logic (shared + app bundles) but also doing the part i mentioned about (renaming the module id's of things like ```crypto-browserify``` to ```crypto```)

package.json
```
...
"dependencies":{
"crypto-browserify":"latest"
}
...
```

gulpfile.ts
```
...
/**
 *    // get all node components ids and use 'resolve' to resolve
  // the ids to their full path, which we need for require()
 */
function getNPMPackageIds(baseDir: string) {
    try {
        // read package.json and get dependencies' package ids
        var packageManifest = require(path.normalize(path.join(baseDir, './package.json')));
        //var xlibPackageManifest = require(path.normalize(path.join(rootPath, "../xlib", './package.json')));

        var packageKeys = _.keys(packageManifest.dependencies);
        //var xlibKeys = _.keys(xlibPackageManifest.dependencies);
        return packageKeys;//.concat(xlibKeys);
    } catch (e) {
        console.log("!!!!!!!!!  missing package.json! in " + baseDir);
        // does not have a package.json manifest
        return [];
    }
}
/**
 *  hack workaround for browserify inability to shared-bundle built-in modules.  see https://github.com/substack/node-browserify/issues/1467#issuecomment-167901918
 *  these modules are installed in the blib project, but need to have their module-id's changed so that browserify doesn't inject it into the app bundle.
 */
function getFinalBlibPackageId(id: string):string {
    //console.log("\tBLIBS= \t" + id);
    switch (id) {
        case "crypto-browserify":
            return "crypto";
            break;
        case "path-browserify":
            return "path";
            break;
        case "stream-browserify":
            return "stream";
            break;
        case "zlib-browserify":
            return "zlib";
            break;
        case "http-browserify":
            return "http";
            break;
        case "https-browserify":
            return "https";
            break;
        case "os-browserify":
            return "os";
            break;
        case "timers-browserify":
            return "timers";
            break;
        default:
            return id;
    }
}

var packages = getNPMPackageIds(".");
var vendorLibrariesToExclude: string[] = [];

    _.forEach(packages , (id) => {
        var resolvedPath = nodeResolve.sync(id);
        id = getFinalPackageId(id);

        b.require(resolvedPath, {
            // exposes the package id, so that we can require() from our code.
            // for eg:
            // require('./vendor/angular/angular.js', {expose: 'angular'}) enables require('angular');
            // for more information: https://github.com/substack/node-browserify#brequirefile-opts
            expose: id,
        });
        vendorLibrariesToExclude.push(id);
    });

///////////////////////////
///////  create shared bundle
var browserifyOptions = {}; //set options to something for the shared.bundle.js
var b = browserify(browserifyOptions);
b.bundle();

///////////////////////////
//////// create app bundle
var browserifyOptions = {}; //set options to something for the app.bundle.js
var b = browserify(browserifyOptions);
//////////////////////////////////////
// specify our vendor bundled modules as external
for (var i = 0; i < vendorLibrariesToExclude.length; i++) {
    b.external(vendorLibrariesToExclude[i]);
}
b.bundle();
...
```

**Comment by jmm** [comment]
Hello @jasonswearingen. When @terinjokes suggested

> a gist using just the browserify API

he very likely didn't mean

* a gulpfile 
* something with a bunch of commented-out code
* something written in a non-JS language

and likely meant an actual gist (which can be cloned, though if it were me I'd suggest a regular repo). I'm mentioning this because an example like this makes it more work for someone to help you (and therefore makes it less likely someone will). Providing something minimal and focused is much more conducive to anything happening.

It's besides the point, but why would you use a `switch` instead of a simple map POJO for those module names?

**Comment by jasonswearingen** [comment]
@jmm 
1) there are 3 lines of commented code.  the rest is documentation.  fyi i cleaned this up quite a bit.
2) if you look at the code you'll see i actually am just dealing with browserify functionality.   nothing in there is gulp related.
3) you'll be hard pressed to see anything non-JS-language in that code.    a couple ```id:string``` like things that are pretty self explanatory.

This should be very easy for a browserify contributor understand the workaround....   if not, well sorry I gave it my best shot.  sorry i only work with browserify via gulp, and my actual gulpfile is very very complicated.

re ```switch```, a POJO would be slightly better, but hindsight 20/20 I suppose.

**Comment by terinjokes** [comment]
@jasonswearingen woah, why are you recursing over all of your package's dependencies and requiring them? You should probably simply require your entry file (likely something like "index.js" or "entry.js" or "app.js"), and browserify would automatically resolve any dependencies you have (and you can still exclude whatever you're considering to be vendor libraries).

**Comment by terinjokes** [comment]
@jmm i like gists, they don't clutter up my repositories list, so they tend to stick around longer (in case someone down the road ends up wanting to review). But personal preference, and I'd happily work with either :smile_cat:

**Comment by jmm** [comment]
@terinjokes I've had a hard time warming up to them for some reason. One thing I don't like is that it doesn't seem to be possible to create an empty one and just push the content to it. For people providing repros for most non-trivial things it'd definitely be better than embedding the code in a comment, but with the seeming emphasis on entering the file content via a web form I wonder how people do with making repros that actually run.

**Comment by jasonswearingen** [comment]
@terinjokes ,   not recursing, but looping (i can still see why that would in general be bad)

This "find all 3rd party modules" logic is part of the multi-bundle workflow.  it's so all those modules can be found so they can be put in the ```shared-vendor.bundle.js```  and then excluded from ```app.bundle.js```  

sorry again about not including a gist.  never used them, nor browserify by itself, so creating one for this would take many many hours :(

PS: I see the "multiple bundles" link I posted earlier is actually for creating multiple app bundles at the same time.   totally ***not*** related to creating shared-vendor bundles!  here's the right link to the tutorial I followed to create shared bundles:  https://github.com/sogko/gulp-recipes/tree/master/browserify-separating-app-and-vendor-bundles  @terinjokes this is where I got the ```getNPMPackageIds()``` function from.
